mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp

```diff
@@ -675,16 +675,67 @@ Value mlir::rock::insertTransposeAndBroadcastTransforms(
     LLVM_DEBUG(llvm::dbgs() << "Reached makeBroadcast with map " << inpIdxMap
                             << " and diff = " << diff << "\n");
 
+    if (diff < 0) {
+      // collapse non-dim exprs
+      // inp = rock.transform(inp) {[0, 1], 2, 3}
+      MutableAffineMap newInpIdxMap = AffineMap::getMinorIdentityMap(
+          outShape.size(), outShape.size(), b.getContext());
+      uint32_t newIdx = 0;
+      SmallVector<SmallVector<uint32_t>> merges;
+      SmallVector<uint32_t> mergeDims;
+      for (const auto &idxAndValue : llvm::enumerate(inpIdxMap.getResults())) {
+        uint32_t idx = idxAndValue.index();
+        AffineExpr resultExpr = idxAndValue.value();
+        mergeDims.push_back(idx);
+        if (diff != 0 && resultExpr.isa<AffineConstantExpr>() &&
+            inpShape[idx] == 1) {
+          diff++;
+        } else {
+          newInpIdxMap.setResult(newIdx++, resultExpr);
+          merges.push_back(mergeDims);
+          mergeDims.clear();
+        }
+      }
+      if (mergeDims.size())
+        merges.back().append(mergeDims);
+
+      TopDownTMBuilder collapseTransform(b, outShape, loc);
+      for (auto idxAndMerge : llvm::enumerate(merges)) {
```

Comment:
nit : this feels like a good utility for transform map utils to create reassociation transforms that might be useful elsewhere w/o having to deal with names for axes.

Reply:
So the suggestion was to lift this as a function here...